src/librustc_typeck/check/pat.rs

```diff
@@ -411,22 +401,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let msg = |ty| {
             format!("this is of type `{}` but it should be `char` or numeric", ty)
         };
-        let mut one_side_err = |first_span, first_ty, second_span, second_ty: Ty<'_>| {
+        let mut one_side_err = |first_span, first_ty, second: Option<(bool, Ty<'tcx>, Span)>| {
             err.span_label(first_span, &msg(first_ty));
-            if !second_ty.references_error() {
-                err.span_label(
-                    second_span,
-                    &format!("this is of type `{}`", second_ty)
-                );
+            match second {
+                Some((_, ty, sp)) if !ty.references_error() => {
+                    err.span_label(sp, &format!("this is of type `{}`", ty));
+                }
+                _ => {}
             }
         };
-        if lhs_fail && rhs_fail {
-            err.span_label(begin_span, &msg(lhs_ty));
-            err.span_label(end_span, &msg(rhs_ty));
-        } else if lhs_fail {
-            one_side_err(begin_span, lhs_ty, end_span, rhs_ty);
-        } else {
-            one_side_err(end_span, rhs_ty, begin_span, lhs_ty);
+        match (lhs, rhs) {
+            (Some((true, lhs_ty, lhs_sp)), Some((true, rhs_ty, rhs_sp))) => {
+                err.span_label(lhs_sp, &msg(lhs_ty));
+                err.span_label(rhs_sp, &msg(rhs_ty));
+            }
+            (Some((true, lhs_ty, lhs_sp)), rhs) => one_side_err(lhs_sp, lhs_ty, rhs),
+            (lhs, Some((true, rhs_ty, rhs_sp))) => one_side_err(rhs_sp, rhs_ty, lhs),
+            _ => {} // Impossible, verified above.
```

Comment:
```suggestion
            _ => span_bug!(span, "Impossible, verified above."),
```

is just as documenting, but with code